Source/WebKit/UIProcess/WebProcessPool.messages.in

```diff
@@ -20,16 +20,17 @@
 # OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
+[SharedPreferencesNeedsConnection]
 messages -> WebProcessPool {
     HandleMessage(String messageName, WebKit::UserData messageBody)
     HandleSynchronousMessage(String messageName, WebKit::UserData messageBody) -> (WebKit::UserData returnData) Synchronous
 
 #if ENABLE(GAMEPAD)
     // FIXME: Consider moving Gamepad messages to their own MessageListener.
-    [EnabledIf='sharedPreferencesForWebProcess(connection).gamepadsEnabled'] StartedUsingGamepads()
-    [EnabledIf='sharedPreferencesForWebProcess(connection).gamepadsEnabled'] StoppedUsingGamepads() -> ()
-    [EnabledIf='sharedPreferencesForWebProcess(connection).gamepadsEnabled'] PlayGamepadEffect(unsigned gamepadIndex, String gamepadID, enum:uint8_t WebCore::GamepadHapticEffectType type, struct WebCore::GamepadEffectParameters parameters) -> (bool success)
-    [EnabledIf='sharedPreferencesForWebProcess(connection).gamepadsEnabled'] StopGamepadEffects(unsigned gamepadIndex, String gamepadID) -> ()
+    [EnabledBy=GamepadsEnabled] StartedUsingGamepads()
```

Comment:
The code is adding a mapping from connection to web process for each message.

This use case is mapping gaming message gamepads to web processes that are allowed to send the gamepad messages.

I think both would be better solved by moving the messages to a dedicated message listener.

E.g. using what's called "capability based access control".
* The access would  checked when the message listener is created.  E.g. caller sends message "create gamepad manager id 1", and then when handling this message, the preference would be checked.)
* If the listener exists, it is a proof that the sender is allowed to do the operation 

This is on the grounds of the capability based access control is already used, as it's the more general method and needed everywhere. In other words, everywhere we need to split up the complex functionality to submodules (like audio destination managers, graphics contexts, etc. and not everything is added to a single massive interface such as WebProcessPool ). 


Additional bug seems to be that StartedUsingGamepads are added to WebProcessPool instead of WebProcess.
Current logic:
* WebProcess A: send a message to WebProcessPool
* UI: Look up WebProcessProxy of the sender based on IPC::Connection
* UI: Look up preferences of the WebProcessProxy

More sane logic:
* WebProcess A: send a message to WebProcessProxy
* UI: Look up preferences of the WebProcessProxy

Same thing for the gamepad logic: it's pointless to send to WebProcessPool if the first thing the code does is to look up WebProcessProxy.

Reply:
I've discussed that design but the gamepads state need to be shared across all web processes so we're keeping the current design.